[ATAPI] Add ATA storage driver

disean opened the pull request:
## Purpose

This patch introduces a new PnP-aware ATA storage stack for PATA and AHCI devices. The implementation differs from Windows one, but the storprop interface is NT6+ compatible. The old ATA driver (UniATA) can be enabled via a CMake define (-DENABLE_UNIATA=1). The stack consists of the following modules:
- pciide.sys

This driver helps maintaining backwards compatibility with NT5.x and its only purpose is to load the pciidex driver for an ATA controller.

- atapi.sys

This driver implements an ATA port driver that communicates with class drivers and the storprop.dll library. The driver translates SCSI commands into the appropriate ATA commands for attached devices and implements ATA-specific error handling.

- pciidex.sys

This driver provides a low-level interface to the ATA bus. The interface allows the port driver to send ATA commands and configure the transfer speed. Current supported ATA controllers: Generic PATA and AHCI controllers, AMD, ATI, CMD, Intel, SIL680, ServerWorks, Toshiba, VIA, PC-98 internal IDE interface. More controllers could be added later if needed.

- legide.sys

This driver detects non-PnP PATA controllers in the system. The legacy PATA device detection should done after PCI bus enumeration, therefore this logic is implemented in a separated kernel-mode driver. MS dropped support for it in Windows Vista.

- storprop.dll

The library should be used to display information about ATA devices to the user. However, the property page is currently not implemented.

JIRA issue: [CORE-17256](https://jira.reactos.org/browse/CORE-17256)

https://reactos.org/testman/compare.php?ids=106659,106663,106667

disean committed:
[NDK] Introduce function attributes for section placement

Increase code readability by using the function attributes
instead of a plain DECLSPEC_NOINLINE

disean committed:
[DDK] Add ATA definitions

Based on the MinGW header
CORE-17256

disean committed:
[PCIIDEX] Add DMA support

CORE-17256

disean committed:
[ATAPI] Add ATA storage driver

CORE-17256
CORE-17191
CORE-17716
CORE-17977
CORE-13976
CORE-16216

binarymaster commented on drivers/storage/ide/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 
 add_subdirectory(atapi)
 add_subdirectory(pciide)
+add_subdirectory(pciahci)
```
Alphabetically it should be above `pciide`.

binarymaster commented on sdk/include/ndk/section_attribs.h:
```diff
@@ -33,3 +33,20 @@ Header Name:
+/*
+ * This attribute should be applied to a function that
+ * called from pageable code and raises the IRQL at DISPATCH_LEVEL or higher,
+ * and restores the IRQL before it returs. We must make sure that function is not inlined.
```
```suggestion
 * and restores the IRQL before it returns. We must make sure that function is not inlined.
```

binarymaster commented on media/inf/hdc.inf:
```diff
@@ -310,36 +310,42 @@ HKR, , Installer32,    0, "syssetup.dll,HdcClassInstaller"
-%*PNP0600.SecDeviceDesc%=uniata_Inst, Secondary_IDE_Channel
-%*PNP0600.DeviceDesc%=uniata_Inst, *PNP0600                            ; Generic_ESDI_Hard_Disk_Controller   (BAD_IDE)
-%UniATA_Name%=uniata_Inst
+%PCI\CC_0104.DeviceDesc%=PciIde_Inst,PCI\CC_0104                       ; Device ID for generic Dual PCI IDE  (UNKNOWN_PCI_IDE)
+%PCI\CC_0105.DeviceDesc%=PciIde_Inst,PCI\CC_0105                       ; Device ID for generic Dual PCI IDE  (UNKNOWN_PCI_IDE)
+%PCI\CC_0106.DeviceDesc%=PciAhci_Inst,PCI\CC_0106                      ; Device ID for generic Dual PCI IDE  (UNKNOWN_PCI_IDE)
+%*PNP0600.PriDeviceDesc%=Atapi_Inst, Primary_IDE_Channel
+%*PNP0600.SecDeviceDesc%=Atapi_Inst, Secondary_IDE_Channel
+%*PNP0600.DeviceDesc%=Atapi_Inst, *PNP0600                            ; Generic_ESDI_Hard_Disk_Controller   (BAD_IDE)
+%UniATA_Name%=PciIde_Inst
```
Just a thought, maybe it would be better to move from global hdc.inf to an inf file within the driver module.

binarymaster reviewed (commented):
For the commit message:
```
[ATAPI] Remove the driver 

This driver was originally added as part of 4e7b22b
Remove the driver as it's been replaced
with UniATA driver in 8eab3a3
```

I suggest specifying SVN revisions for pre-git converted commits this way: `4e7b22b (r26238), 
8eab3a3 (r29286)` etc.

Also if the removal of old ATAPI driver doesn't break anything (it's unused, right?), it could be done in a separate janitorial PR.

--------------

Nit suggestion for the headers in new files, please use `<name@email>` triangle brackets.

disean committed:
[UNIATA] Disable the driver and enable the new ATA stack

Also fix descriptions for devices 1166:0241 and 1166:0242.
See https://bugzilla.kernel.org/show_bug.cgi?id=10424 and
https://github.com/torvalds/linux/commit/aeb74914ef0e6746f15b11c0399048de9c5c05b8

CORE-17256

disean committed:
[ATAPI] Disable the driver

binarymaster commented on drivers/storage/ide/atapi/CMakeLists.txt:
```diff
@@ -1,6 +1,36 @@
-add_library(atapi MODULE atapi.c atapi.rc)
-target_link_libraries(atapi libcntpr)
+if(ARCH STREQUAL "i386")
+  set(DETECT_LEGACY_DEVICES TRUE)
+  add_definitions(-DATA_DETECT_LEGACY_DEVICES)
```
```suggestion
    set(DETECT_LEGACY_DEVICES TRUE)
    add_definitions(-DATA_DETECT_LEGACY_DEVICES)
```

binarymaster commented on drivers/storage/ide/atapi/atapi.h:
```diff
@@ -1,461 +1,887 @@
+
+#define ATA_MAX_LUN_COUNT        8
+
+/* Maxumum size (ATA Information VPD page) */
```
```suggestion
/* Maximum size (ATA Information VPD page) */
```

JoachimHenze commented on drivers/storage/ide/atapi/atahw.h:
```diff
@@ -0,0 +1,654 @@
+ATA_READ(
+  _In_ PUCHAR Port)
+{
+    return READ_PORT_UCHAR(Port);;
```
```suggestion
    return READ_PORT_UCHAR(Port);
```

JoachimHenze commented on drivers/storage/ide/atapi/debug.h:
```diff
@@ -0,0 +1,72 @@
+#define WARN
+#define ERR
+
+#endif
```
What's your reason for redefining your set of debug-macros instead of reusing existing ones we have?

disean replied:
I want to limit the amount of debug information

JoachimHenze replied:
I would have expected that to be also possible using DPRINT/DPRINT1, but ok. Don't want to stress that. PR is large.

JoachimHenze commented on drivers/storage/ide/atapi/hw.c:
```diff
@@ -0,0 +1,878 @@
+    ATA_WRITE(IoBase + DMA_COMMAND, DMA_COMMAND_STOP);
+    ATA_WRITE(IoBase + DMA_STATUS, DMA_STATUS_INTERRUPT | DMA_STATUS_ERROR);
+
+    /* Beginning of the the PRD Table */
```
What's a PRD table? What does the abbreviation stand for? **P**CI **R**esource **D**escriptor? Or anything else?

disean replied:
Physical Region Descriptor Table. A table that lists the scatter/gather elements.

JoachimHenze commented on drivers/storage/ide/atapi/hw.c:
```diff
@@ -0,0 +1,878 @@
+
+    /*
+     * For PIO writes we need to send the first data block
+     * to make the devive to start generating interrupts.
```
```suggestion
     * to make the device start generating interrupts.
```

JoachimHenze commented on drivers/storage/ide/atapi/satl.c:
```diff
@@ -0,0 +1,1459 @@
+    /* Choose the command opcode */
+    Request->TaskFile.Command = AtaReadWriteCommand(Request, DeviceExtension);
+
+    /* Terminate the request when the FUA is not supported */
```
What is a FUA?

disean replied:
https://en.wikipedia.org/wiki/Disk_buffer#FUA

JoachimHenze commented on drivers/storage/ide/atapi/satl.c:
```diff
@@ -0,0 +1,1459 @@
+            UNREACHABLE;
+    }
+
+    /* Make sure the sector count do not exceed the maximum transfer length */
```
```suggestion
    /* Sector count must not exceed the maximum transfer length */
```

JoachimHenze commented on drivers/storage/ide/pciidex/pciidex.c:
```diff
@@ -239,6 +250,12 @@ PciIdeXAddDevice(
 
+    if (FdoExtension->Flags & FDO_AHCI)
+    {
+        DPRINT1("ACHI support not implemented yet\n");
```
```suggestion
        DPRINT1("AHCI support not implemented yet\n");
```

JoachimHenze commented on drivers/storage/ide/atapi/hw.c:
```diff
@@ -0,0 +1,878 @@
+    ATA_WRITE(IoBase + DMA_COMMAND, DMA_COMMAND_STOP);
+    ATA_WRITE(IoBase + DMA_STATUS, DMA_STATUS_INTERRUPT | DMA_STATUS_ERROR);
+
+    /* Beginning of the the PRD Table */
```
```suggestion
    /* Beginning of the the Physical Region Descriptor (PRD) Table */
```
Might make sense to write it out at least once at one location where it is already used, e.g. here.

binarymaster commented on drivers/storage/ide/atapi/data.c:
```diff
@@ -0,0 +1,88 @@
+    {"Other",      "GenOther",      "OtherPeripheral"          }, // 10
+};
+
+const CHAR* const AtapBrokenInquiryDrive[] =
+{
+    "THOMSON-DVD",
+    "PHILIPS XBOX DVD DRIVE",
+    "PHILIPS J5 3235C",
+    "SAMSUNG DVD-ROM SDG-605B"
+};
```
```suggestion
const CHAR* const AtapBrokenInquiryDrive[] =
{
    "THOMSON-DVD",
    "PHILIPS XBOX DVD DRIVE",
    "PHILIPS J5 3235C",
    "SAMSUNG DVD-ROM SDG-605B",
    /* TODO: Xbox 360 drives are also affected. */
};
```
Technically it's possible to connect Xbox DVD drives to an ordinary PC.

Regarding X360 drives, it was mentioned somewhere in XboxDev.

disean replied:
I failed to find anything useful in the wiki. Can you list the X360 drives here please?

binarymaster replied:
It can be postponed then, since I have no information for these.

binarymaster commented on boot/bootdata/txtsetup.sif:
```diff
@@ -90,7 +90,7 @@ FreeSysPartDiskSpace=450
 nmidebug.sys = 1,,,,,,x,4,,,,1,4
 sacdrv.sys   = 1,,,,,,x,4,,,,1,4
-uniata.sys   = 1,,,,,,x,4,,,,1,4
+atapi.sys   = 1,,,,,,x,4,,,,1,4
```
```suggestion
atapi.sys    = 1,,,,,,x,4,,,,1,4
```

JoachimHenze replied:
This aligned list is a good example by the way why sticking with 8.3 names (whereever possible) is a nice thing to have, @learn-more

binarymaster commented on drivers/storage/ide/atapi/hwpata.c:
```diff
@@ -0,0 +1,877 @@
+
+    /*
+     * For PIO writes we need to send the first data block
+     * to make the devive start generating interrupts.
```
device or drive?

binarymaster replied:
It became "devive" again on one of force pushes 😁

disean committed:
[HALX86:DMA] HACK: Avoid excessive stack usage

CORE-17256

oleg-dubinskiy commented:
Looks OK in general, but one small suggestion from me: please remove (not simply disable) old atapi driver and uniata as well in a separate janitorial PRs (in case you did not do that yet), as it was already suggested above.
> The patch is WIP

Is it still WIP currently? Maybe it's simply not finished on 100% yet? Some unimplemented parts do not mean the PR is WIP and very incomplete, especially if the new driver is working correctly in the most of cases (it was tested many times already by Justin and others). :slightly_smiling_face:

disean commented:
> I'll mark it as ready for review

The driver is not ready for merge, it's still work in progress. I've tested it successfully on real hardware and in an emulation environment. What is currently missing is an error recovery mechanism and support for hardware (hotplug, PATA timings). Sorry for being so inactive lately. I've been working on this driver on my own local copy for the whole time. I prefer to keep the public branch to code that compiles without errors and works on real hardware as possible, so that other people can compile and test it as well.

> please remove (not simply disable) old atapi driver and uniata as well

Maybe we could make UniATA the default fallback for the new ATAPI?

oleg-dubinskiy commented:
>Maybe we could make UniATA the default fallback for the new ATAPI?

Indeed. It may be even a better option, until our new ATAPI is properly functioning in all cases (works on all hardware). :+1:

oleg-dubinskiy commented:
>The driver is not ready for merge, it's still work in progress.

OK, if so, then I'll convert it back to the Draft state.

tkreuzer commented:
Doesn't seem to work with any VBox controller.

binarymaster commented on drivers/storage/ide/atapi/acpi.c:
```diff
@@ -0,0 +1,357 @@
+ * PROJECT:     ReactOS ATA Port Driver
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     ACPI interface with SATA ports, IDE controllers and drives
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
```
You can set copyright years as "2024-2025" on files that you started working on in 2024 🙂

tkreuzer commented:
Crashes on x64: https://build.reactos.org/#/builders/12/builds/6235

disean commented:
I did some debugging around amd64 BootCD. It uses ACPI HAL and ends up causing trouble for acpi.sys to initialize. Hence pci.sys cannot be loaded, the dependent device drivers will never be installed. I applied a workaround to force the installation of the atapi driver, but the APIC interrupt delivery appeared to be non-functional. It's very strange because I've successfully run LiveCD with the atapi driver on amd64 real hardware.

disean commented:
Sorry for the lack of information. I found out what the problem is. For some reason, when boot-time drivers are unable to create a boot device driver object, acpi.sys never gets loaded. This only happens on amd64 BootCD. I've just pushed a workaround to report the storage controllers early (NT4-style device detection), everything works after that. Could you please try again? 

On VMware (amd64 BootCD or amd64 LiveCD), I don't receive interrupts from SATA devices:
```
AtaPortTimeout PORT 0: *Slot 0 (00000001) timed out 618040
AtaPortCompleteInternalRequest Internal request failed 9
```

DarkFire01 commented:
The ACPI fixes lately have made it so you shouldn't need that hack anymore. On top of this I've noticed a massive improvement in the stability of x64 with this driver now :).

There will still be some interrupt issues but they won't show up on the buildbots anymore, and of course vbox still works :).
Please feel free to continue forward with this, as my ACPI + HAL + PCI fixes will improve the interrupt situation. But they'll also make the experience with UNIATA act more like vista (e.g. not work well) so this likely should be merged first.

<img width="789" height="667" alt="image" src="https://github.com/user-attachments/assets/ed3b6377-7ad3-4c80-8b3a-bdf2146e0b80" />

binarymaster commented on drivers/storage/ide/pciidex/acpi.c:
```diff
@@ -0,0 +1,337 @@
+    if (!InputBuffer)
+    {
+        ERR("Failed to allocate memory\n");
+        return Status;
```
`error C4700: uninitialized local variable 'Status' used`

tkreuzer commented:
Works on x64 now. https://reactos.org/testman/compare.php?ids=104166,104194
The missing tests did actually run and succeed. Only urlmon:protocol crashed, but that is very likely random.

tkreuzer commented on drivers/storage/ide/atapi/include/debug.h:
```diff
@@ -0,0 +1,103 @@
+    } while (0)
+
+#else
+#define TRACE
```
Shouldn't this rather be
```suggestion
#define TRACE(fmt, ...)
```

disean replied:
Fixed.

tkreuzer commented on drivers/storage/ide/atapi/atapi.c:
```diff
@@ -1,6539 +1,713 @@
+CODE_SEG("PAGE")
+NTSTATUS
+AtaOpenRegistryKey(
+    _In_ PHANDLE KeyHandle,
```
`_Out_`

tkreuzer commented on drivers/storage/ide/atapi/atapi.c:
```diff
@@ -1,6539 +1,713 @@
+    if (!NT_SUCCESS(Status))
+    {
+        TRACE("Failed to open device hardware key, status 0x%lx\n", Status);
+        return;
```
Use `ERR` here? Also shouldn't the function return the status?

tkreuzer commented on drivers/storage/ide/atapi/data.c:
```diff
@@ -0,0 +1,68 @@
+
+        DEFAULT_UNREACHABLE;
+    }
+}
```
```suggestion
    ASSERT(FALSE);
}
```

tkreuzer commented on drivers/storage/ide/atapi/dev_config.c:
```diff
@@ -0,0 +1,404 @@
+    if (Mode >= UDMA_MODE(0))
+    {
+        Mode -= UDMA_MODE(0);
+        Mode |= 0x40;
```
Where do these magic values come from? I'd appreciate either symbolic constants or a comment / reference.

tkreuzer commented on drivers/storage/ide/atapi/dev_identify.c:
```diff
@@ -0,0 +1,359 @@
+    if (!RtlEqualMemory(IdentifyData1->FirmwareRevision,
+                        IdentifyData2->FirmwareRevision,
+                        RTL_FIELD_SIZE(IDENTIFY_DEVICE_DATA, FirmwareRevision) +
+                        RTL_FIELD_SIZE(IDENTIFY_DEVICE_DATA, ModelNumber)))
```
Comparing 2 fields in one go can lead to static analysis errors. Might be worth to either use a structure that wraps `FirmwareRevision` and `ModelNumber` or do 2 seperate calls to `RtlEqualMemory`.

disean replied:
Ok, fixed.

tkreuzer commented on drivers/storage/ide/atapi/enum.c:
```diff
@@ -0,0 +1,656 @@
+        Destination[i + 1] = Source[i];
+    }
+
+    return &Destination[i];
```
Is this supposed to point at the end (i.e. *after* the buffer)? What's the point of this? It's just gonna be used wrongly (like in `AtaDeviceIsXboxDrive`) and cause a buffer overrun.

tkreuzer commented on drivers/storage/ide/atapi/dev_identify.c:
```diff
@@ -0,0 +1,359 @@
+    End = AtaCopyIdStringUnsafe((PUCHAR)ModelNumber,
+                                DevExt->IdentifyDeviceData.ModelNumber,
+                                sizeof(ModelNumber));
+    *End = ANSI_NULL;
```
This looks like a buffer overrun. If the goal is to terminate the buffer, this should probably be done by `AtaCopyIdStringUnsafe` or do it here as `ModelNumber[25] = ANSI_NULL`.
And should this is really be `ANSI_NULL`, when `AtaCopyIdStringUnsafe` shuffles around `USHORT`s?

disean replied:
Shoud be fixed now.

tkreuzer commented:
Bugcheck reported on hardware: https://chat.reactos.org/reactos/pl/gqezi74j338g88bj6m18bo9mre

binarymaster commented:
From the recent tests I've noticed that the `ATA Channel` device nodes are not numbered anymore. Is this intentional or something to fix?

DarkFire01 commented:
I've gone through most of my devices now and everything with an Intel chipset has worked perfectly.

However AMD Chipsets with SATA seemed to have regressed pretty bad. I saw the timeouts in this log and thought it may have something to do with a lack of arbiter but this is a legacy HAL and testing with Windows 2003 against this driver caused the same thing to happen. Pretty much it'll spin till it gives up and we continue booting (7B bugcheck if its installed to drive, and no disks detected on USB boot)
[debuglog.txt](https://github.com/user-attachments/files/23431959/debuglog.txt)

disean commented:
> [debuglog.txt](https://github.com/user-attachments/files/23431959/debuglog.txt)

This strange problem was caused by the AHCI interrupts not being generated. Presuming that the hardware is sane, it will a bit hard to understand what happening.

DarkFire01 commented:
> > > [debuglog.txt](https://github.com/user-attachments/files/23431959/debuglog.txt)
> > 
> > 
> > This strange problem was caused by the AHCI interrupts not being generated. Presuming that the hardware is sane, it will a bit hard to understand what happening.
> 
> Take a look at the AMD Simnow emulator. You'll be able to reproduce this there I'm pretty sure. but i can check for you if you'd like first :)

a bit of an update,
it seems this happens on pretty much all AMD systems, even later ryzen chipsets.
![AGC_20251118_203000530](https://github.com/user-attachments/assets/f511f32c-8071-4760-ab48-206d4f3facbc)

disean commented:
> [debuglog.txt](https://github.com/user-attachments/files/23431959/debuglog.txt)

@DarkFire01 Can you please try this debug patch and grab a log? 77c7c0eb12c31fbefe458395f96183fdf8bf7161

disean commented:
>  As you can imagine this loops for awhile

Thanks, I've fixed a bug and disabled the PMP support code for this chip. Hope that helps. https://github.com/reactos/reactos/pull/6577/commits/892f599320fb0f66544685a215759596e45df4a9

obsolete-is-better commented:
I'm running into a assert when testing this pr's artifact livecd with Bochs.
[bochs_latest.txt](https://github.com/user-attachments/files/23754766/bochs_latest.txt)

disean commented on drivers/storage/ide/atapi/atapi.c:
```diff
@@ -1,6539 +1,713 @@
+    if (!Irp)
+    {
+        ERR("IoBuildDeviceIoControlRequest() failed\n");
+        return;
```
Fixed.

disean commented on drivers/storage/ide/atapi/dev_config.c:
```diff
@@ -0,0 +1,404 @@
+    PATA_DEVICE_REQUEST Request = &PortData->Worker.InternalRequest;
+
+    Request->Flags = 0;
+    Request->TimeOut = 3;
```
This is a sane value.

disean commented on drivers/storage/ide/atapi/include/debug.h:
```diff
@@ -0,0 +1,103 @@
+#include <string.h>
+#define __FILENAME__ (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
+
+#define My_DbgPrint DbgPrint
```
Some temporary debugging code, to be removed later.

disean commented on drivers/storage/ide/atapi/enum.c:
```diff
@@ -0,0 +1,656 @@
+    _Inout_updates_bytes_(Length * sizeof(USHORT)) PVOID Buffer,
+    _In_range_(>, 0) ULONG Length)
+{
+    PUSHORT Word = Buffer;
```
Added an ASSERT check.

disean commented on drivers/storage/ide/atapi/enum.c:
```diff
@@ -0,0 +1,656 @@
+VOID
+AtaSwapIdString(
+    _Inout_updates_bytes_(Length * sizeof(USHORT)) PVOID Buffer,
+    _In_range_(>, 0) ULONG Length)
```
Renamed to `WordCount` to reflect the number of elements.

disean commented on drivers/storage/ide/atapi/CMakeLists.txt:
```diff
@@ -1,6 +1,48 @@
-add_library(atapi MODULE atapi.c atapi.rc)
-target_link_libraries(atapi libcntpr)
+include_directories(
+    ${REACTOS_SOURCE_DIR}/sdk/lib/drivers/sptilib)
```
I don't know what to do with the sptilib target. It would be nice to provide some examples here.

DarkFire01 commented:
> > > Thanks! Here's the log, sadly still drives aren't detected
> > > [atapilog.txt](https://github.com/user-attachments/files/23754774/atapilog.txt)
> > 
> > 
> > I'm afraid I've no idea how to debug this case any further, the log doesn't give me any further clues. It detects attaches devices correctly, but the IDENTIFY DEVICE command fails after 10 seconds due to timeout. I suspect the issue is coming from somewhere besides the driver code. What kind of mobo is it? I thought it would be helpful to acquire one.
> 
> to be honest this is happening on every single one of my AMD systems, all the way from phoenom to Ryzen.
> 
> so I imagine picking up any AMD system will work, but can give you a list of motherboard I use in my testing suite if you’d like that all did this.
> 
> Appending this list as i go through them today off and on: ROG STRIX B850-A Gaming WIFI MSI 770-C45
> 
> laptops: Compaq CQ62-220US

Reattempting this PR, the machines all listed above work and install now.
Thanks for looking into it!!!

HBelusca commented on sdk/include/ndk/section_attribs.h:
```diff
@@ -33,3 +33,20 @@ Header Name:
 #endif
+
+/*
+ * This attribute should be applied to a function that
```
"to a function that _**is**_ called from pageable code" ?

DarkFire01 reviewed (approved):
yolo

HBelusca commented on drivers/storage/ide/atapi/CMakeLists.txt:
```diff
@@ -1,6 +1,36 @@
+endif()
+
 set_module_type(atapi kernelmodedriver)
-add_importlibs(atapi scsiport)
```
Side-remark: on Windows 7 (and maybe Vista, and probably 8+), MS split the atapi driver into atapi ("miniport") + ataport:
<img width="800" height="440" alt="image" src="https://github.com/user-attachments/assets/68a202cf-5d5f-49a1-8b1e-36bbd76b1318" />

(And what's consumed by the "atapi" miniport from ntoskrnl, is only KeBugCheckEx.)

Maybe in the future we'll find useful to perform a similar split.

HBelusca reviewed (commented):
Atapi driver reviewed (1/7)

HBelusca reviewed (commented):
LegIDE / PciIDE review (2/7)

HBelusca commented on drivers/storage/ide/pciidex/pata.h:
```diff
@@ -0,0 +1,304 @@
+    _In_ UCHAR DeviceSelect)
+{
+#if defined(_M_IX86)
+    /* NEC extension to allow 4 drives per channel */
```
Doesn't need a `if (IsNEC_98)` check? Or the intent is that is supported on **all** NEC drives that can be used on an x86 platform, whether it is PC98 or not?

disean replied:
No. PC-98 systems have a PCI bus and you can plug a PCI IDE controller into a PCI bus slot. https://github.com/disean/reactos/blob/21f80686bc6fa7e860086d75c0d01f5429ec12cb/drivers/storage/ide/pciidex/chipset/pata_generic.c#L747-L759

HBelusca reviewed (commented):
PciIdeX review (3/7)

HBelusca reviewed (commented):
PciIdeX review (4/7)

HBelusca reviewed (commented):
Misc files review (5/7)

HBelusca reviewed (commented):
Review finished (parts 6 and 7/7).

As previously mentioned in some comment, it would be preferable to do a separate commit that moves the original storage/ide/atapi to storage/ide/atapi_old , so as to get the commit history clean, _**and then**_, apply the whole PR.

disean commented on drivers/storage/ide/atapi/atapi.h:
```diff
@@ -1,461 +1,1020 @@
+        __assume(cond); \
+    } while (0)
+
+#define ATAPORT_TAG             'PedI'
```
MS drivers still use `PedI` and `XedI`, which are actually documented in pooltag.txt.

HBelusca replied:
ok then.

disean commented on drivers/storage/ide/atapi/scsi.c:
```diff
@@ -0,0 +1,1810 @@
+     * the requests may take a long period of time to process,
+     * and therefore we have to do the cancellation ourselves.
+     */
+    (VOID)IoSetCancelRoutine(Irp, AtaReqDeviceQueueCancelIo);
```
Nope. I think this is a very large topic:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/drivers/csq/csq.c;hb=7779fb44597a0892c6b1d76f40a6179ca9eccf45#l232

disean commented on drivers/storage/ide/atapi/fdo.c:
```diff
@@ -0,0 +1,751 @@
+        }
+
+        /* Register ourselves (ATA channel) as a SCSI port adapter */
+        ++IoGetConfigurationInformation()->ScsiPortCount;
```
Done

disean commented on drivers/storage/ide/atapi/fdo.c:
```diff
@@ -0,0 +1,751 @@
+{
+    UNICODE_STRING ChannelName;
+    ULONG ScsiAdapter, ScsiPortCount;
+    WCHAR ChannelNameBuffer[sizeof("\\Device\\Ide\\IdePort999")];
```
This limitation is arbitrary. Let's stick to `99999`, since `999` is small for some use cases.

disean commented on drivers/storage/ide/atapi/wmi.c:
```diff
@@ -0,0 +1,213 @@
+    PUNICODE_STRING InstanceName,
+    PUNICODE_STRING *RegistryPath,
+    PUNICODE_STRING MofResourceName,
+    PDEVICE_OBJECT *Pdo)
```
Fixed.

disean commented on drivers/storage/ide/atapi/scsi.c:
```diff
@@ -0,0 +1,1810 @@
+
+static
+VOID
+AtaReqDeviceQueueRemoveEntry(
```
Annotation added.

disean commented on drivers/storage/ide/atapi/scsi.c:
```diff
@@ -0,0 +1,1810 @@
+
+    ASSERT(KeGetCurrentIrql() == DISPATCH_LEVEL);
+
+    if (IsListEmpty(&PortData->PortQueueList))
```
Annotation added.

disean commented on drivers/storage/ide/legide/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+list(APPEND SOURCE
+    legide.c)
```
Done.

disean commented on drivers/storage/ide/legide/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+add_library(legide MODULE ${SOURCE} legide.rc)
+set_module_type(legide kernelmodedriver)
+if(MSVC)
+    # Required for NT5.x pciidex.sys
```
Ok.

disean commented on drivers/storage/ide/legide/legide.c:
```diff
@@ -0,0 +1,853 @@
+/*
+ * This driver detects non-PnP PATA controllers in the system.
+ * The legacy PATA device detection should done after PCI bus enumeration,
+ * therefore this logic is implemented in a separated kernel-mode driver.
```
Fixed.

disean commented on drivers/storage/ide/legide/legide.c:
```diff
@@ -0,0 +1,853 @@
+    if (AtaLegacyIdentifyDevice(Registers, DeviceNumber, IDE_COMMAND_ATAPI_IDENTIFY))
+        return TRUE;
+
+    return FALSE;
```
Done.

disean commented on drivers/storage/ide/pciidex/chipset/ahci_hw.c:
```diff
@@ -0,0 +1,1574 @@
+        return CONN_STATUS_FAILURE;
+    }
+
+    KeStallExecutionProcessor(20);
```
Yes

HBelusca replied:
Maybe leave a comment in the code then.

disean replied:
Done.

disean commented on drivers/storage/ide/pciidex/chipset/pata_io.c:
```diff
@@ -0,0 +1,1217 @@
+            ASSERT((TransferLength % sizeof(USHORT)) == 0);
+
+            /* Make sure we do not write beyond the end of the PRD table */
+            ASSERT(DescriptorNumber++ < ChanData->MaximumPhysicalPages);
```
Fixed.

disean commented on drivers/storage/ide/pciidex/power.c:
```diff
@@ -19,8 +16,6 @@ PciIdeXPdoDispatchPower(
     NTSTATUS Status;
     PIO_STACK_LOCATION IoStack;
 
-    UNREFERENCED_PARAMETER(PdoExtension);
```
Reverted.

disean commented on drivers/storage/ide/pciidex/pciidex.h:
```diff
@@ -115,32 +512,396 @@ DRIVER_UNLOAD PciIdeXUnload;
+
+BOOLEAN
+PciFindDevice(
+    _In_ __inner_callback PATA_PCI_MATCH_FN MatchFunction,
```
They're technically the same thing.

HBelusca replied:
If there's a specific intent you want to communicate with the "inner_callback", then OK, otherwise you may use the shorter  `__callback` annotation.

disean replied:
Replaced with the ` __callback` annotation.

disean commented on drivers/storage/ide/pciidex/pciidex.c:
```diff
@@ -7,11 +7,110 @@
+    _In_ ULONG ConfigDataOffset,
+    _In_ ULONG BufferLength)
+{
+    (*Controller->BusInterface.GetBusData)(Controller->BusInterface.Context,
```
Done.

disean commented on drivers/storage/ide/pciidex/chipset/ahci_generic.c:
```diff
@@ -0,0 +1,753 @@
+        ULONG Index;
+    } AbarLocations[] =
+    {
+        { 0x177D, 0xA01C, 0 },
```
`ThunderX AHCI SATA Controller`

HBelusca replied:
Please add a comment in code to describe each of these (if you don't use named defines for these values).

disean replied:
Done.

disean commented on drivers/storage/ide/atapi/CMakeLists.txt:
```diff
@@ -1,6 +1,36 @@
+endif()
+
 set_module_type(atapi kernelmodedriver)
-add_importlibs(atapi scsiport)
```
In terms of architecture, the `atapi.sys` driver resembles the `ataport.sys` in NT6.x making it easier to perform a split in the future.

disean commented on drivers/storage/ide/atapi/pdo.c:
```diff
@@ -0,0 +1,798 @@
+AtaPdoCompletionRoutine(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_reads_opt_(_Inexpressible_("varies")) PVOID Context)
```
But this is how it's intended to work. https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/xdk/iotypes.h;hb=7779fb44597a0892c6b1d76f40a6179ca9eccf45#l2836

HBelusca replied:
For sure. Here I was actually wondering about the _**usefulness**_ of the annotation in this precise context (if I may say^^).

disean commented on drivers/storage/ide/legide/legide.c:
```diff
@@ -0,0 +1,853 @@
+/*
+ * PROJECT:     Legacy PATA Bus Enumerator Driver
```
It was implemented in NT5.x as a part of `atapi.sys`. MS dropped support for it in Windows Vista. See `Operating system support` https://86box.readthedocs.io/en/latest/hardware/ideterqua.html

HBelusca replied:
I see. So that's why you preferred to move this into a dedicated driver, instead of putting it into the new atapi? (Would make sense from the point of view of being able to test/use the functionality on Windows 2003/Vista+ by driver swapping.)

disean replied:
This won't work properly because when you call pciidex from atapi (ataport) it'll overwrite the driver's dispatch routines (DriverObject->MajorFunction).

disean commented on drivers/storage/ide/atapi/dev_power.c:
```diff
@@ -0,0 +1,531 @@
+    _In_ PATAPORT_DEVICE_EXTENSION DevExt,
+    _Inout_ PIRP Irp)
+{
+    // FIXME: Investigate how to fix this kernel bug
```
This works on a checked build of Windows 2003 (SMP), it looks like a regression introduced by https://github.com/reactos/reactos/pull/7366.

disean commented on drivers/storage/ide/pciidex/chipset/svw_pata.c:
```diff
@@ -0,0 +1,361 @@
+        case PCI_DEV_OSB4_IDE:
+            /* UDMA is supported but its use is discouraged */
+            SupportedMode = PIO_ALL | MWDMA_ALL;
+
```
Why?

disean commented on drivers/storage/ide/pciidex/chipset/ahci_hw.c:
```diff
@@ -0,0 +1,1574 @@
+     * before interacting with the port again,
+     * but on some AHCI controllers this bit never becomes set.
+     */
+    KeStallExecutionProcessor(50);
```
Implemented.

disean commented on drivers/storage/ide/pciidex/chipset/pata_generic.c:
```diff
@@ -0,0 +1,1245 @@
+                                InterruptDesc->u.Interrupt.Level,
+                                (InterruptDesc->Flags & CM_RESOURCE_INTERRUPT_LATCHED)
+                                ? Latched
+                                : LevelSensitive,
```
Done.

disean commented on drivers/storage/ide/pciidex/chipset/intel.c:
```diff
@@ -0,0 +1,1061 @@
+            break;
+        }
+
+        default:
```
Not really, `DEFAULT_UNREACHABLE` doesn't assert anything.

disean commented on drivers/storage/ide/pciidex/chipset/svw_pata.c:
```diff
@@ -0,0 +1,361 @@
+        return FALSE;
+    }
+
+    HalGetBusDataByOffset(PCIConfiguration,
```
This method accesses the PCI southbridge registers while the driver owns the PATA controller. Maybe we should move this to `pci.sys` or to `SYSTEM\CurrentControlSet\Control\HAL`.

disean commented on drivers/storage/ide/pciide/pciide.c:
```diff
@@ -16,59 +16,18 @@ PciIdeChannelEnabled(
-
 	DPRINT("PciIdeChannelEnabled(%p, %lu)\n", DeviceExtension, Channel);
 
-	Status = PciIdeXGetBusData(
```
Yes, since `pciide.sys` is a *generic* IDE controller minidriver.

disean commented on drivers/storage/ide/pciidex/chipset/ahci_generic.c:
```diff
@@ -0,0 +1,753 @@
+{
+    USHORT VendorID;
+    USHORT DeviceID;
+} AhciControllerList[] =
```
The list contains SATA controllers that are known to support AHCI mode. This list could be expanded further to handle controller specific features/errata.

disean commented on drivers/storage/ide/pciidex/pciidex.h:
```diff
@@ -6,47 +6,344 @@
+(CONTROLLER_CHANNEL_ENABLED)(
+    _In_ PATA_CONTROLLER Controller,
+    _In_ ULONG Channel);
+typedef CONTROLLER_CHANNEL_ENABLED *PCONTROLLER_CHANNEL_ENABLED;
```
For example, I need a separate `CHANNEL_CHECK_INTERRUPT` typedef in order to compile the `pata_generic.c` and `pata_io.c` translation units which depend on each other.

disean commented on drivers/storage/ide/pciidex/chipset/svw_sata.c:
```diff
@@ -0,0 +1,283 @@
+VOID
+SvwSataLoadTaskFile(
+    _In_ CHANNEL_DATA_PATA* __restrict ChanData,
+    _In_ ATA_DEVICE_REQUEST* __restrict Request)
```
There are some architectures where `READ_REGISTER_*` and `WRITE_REGISTER_*` are implemented as inline functions. OK, reverted the changes.

HBelusca commented on drivers/storage/ide/pciidex/chipset/svw_sata.c:
```diff
@@ -0,0 +1,283 @@
+    if (!IoBase)
+        return STATUS_DEVICE_CONFIGURATION_ERROR;
+
+    /* Errata fix */
```
Then, please say so explicitly in the code comment.

disean replied:
Comment added.

HBelusca commented on drivers/storage/ide/pciidex/chipset/intel.c:
```diff
@@ -0,0 +1,1061 @@
+ * PURPOSE:     Intel ATA controller minidriver
+ * COPYRIGHT:   Copyright 2026 Dmitry Borisov <di.sean@protonmail.com>
+ *
+ * REFERENCES:  For more details, see Intel order no 298600-004
```
OK then could you please write the whole title of the manual?
```suggestion
 * REFERENCES:  For more details, see: "Intel ICH0~ICH5 Programmer's Reference Manual (PRM) 298600-004"
```

disean replied:
Fixed.

HBelusca commented on media/inf/hdc.inf:
```diff
@@ -574,8 +583,8 @@ PCI\VEN_1166&DEV_0214.DeviceDesc="ServerWorks HT1000 ATA-100 controller"
-PCI\VEN_1166&DEV_0241.DeviceDesc="ServerWorks Frodo4 SATA controller"
-PCI\VEN_1166&DEV_0242.DeviceDesc="ServerWorks Frodo8 SATA controller"
+PCI\VEN_1166&DEV_0241.DeviceDesc="ServerWorks Frodo8 SATA controller"
+PCI\VEN_1166&DEV_0242.DeviceDesc="ServerWorks Frodo4 SATA controller"
```
@disean I would suggest you mention this fact in the commit log then. (So that we keep a trace of that somehow.)

disean commented on sdk/include/reactos/drivers/ata/ntddata.h:
```diff
@@ -0,0 +1,569 @@
+    DEV_NONE = 3
+} ATA_DEVICE_TYPE;
+
+#ifndef NTOS_MODE_USER
```
Added two headers: `ata_shared.h` and `ata_user.h`.

HBelusca reviewed (approved):
Thanks @disean for having taken time to answer all of my questions.
